File: codegen/results/v002/qwen3.5_27b_nothink/20260411_193102/typescript/007_beatles_interview/solution.ts
import * as fs from 'fs';

const REF_DATE = new Date('2025-07-01T00:00:00Z');

interface Relative {
  FirstName: string;
  LastName: string;
  Relationship: string;
}

interface OutputPerson {
  FirstName: string;
  LastName: string;
  Birthday: string;
  Age: number;
  Relatives: Relative[];
}

function parseDate(dateStr: string): Date | null {
  if (dateStr.toLowerCase() === 'null') return null;
  // Format: M/D/YYYY (e.g., 10/9/1940)
  const parts = dateStr.split('/');
  if (parts.length !== 3) return null;
  
  const month = parseInt(parts[0], 10) - 1; // 0-indexed
  const day = parseInt(parts[1], 10);
  const year = parseInt(parts[2], 10);
  
  return new Date(year, month, day);
}

function calculateAge(birthday: Date, refDate: Date): number {
  let age = refDate.getFullYear() - birthday.getFullYear();
  const m = refDate.getMonth() - birthday.getMonth();
  
  if (m < 0 || (m === 0 && refDate.getDate() < birthday.getDate())) {
    age--;
  }
  
  return age;
}

function splitName(name: string): { first: string; last: string } {
  const parts = name.trim().split(/\s+/);
  // Assuming format: "FirstName LastName" or "FirstName MiddleName LastName"
  // We treat the last part as LastName and everything before as FirstName
  // Based on expected output: "John Winston Lennon" -> "John" (FirstName) "Lennon" (LastName)
  // It seems they take the first word as FirstName and the last word as LastName.
  if (parts.length === 2) {
    return { first: parts[0], last: parts[1] };
  } else {
    return { first: parts[0], last: parts[parts.length - 1] };
  }
}

function parseName(name: string): { first: string; last: string } | null {
  if (!name || name.toLowerCase() === 'null') return null;
  return splitName(name);
}

function main() {
  const csvPath = 'input/input.csv';
  
  if (!fs.existsSync(csvPath)) {
    process.stdout.write('[]');
    return;
  }

  const rawData = fs.readFileSync(csvPath, 'utf-8');
  const lines = rawData.split('\n').filter(line => line.trim() !== '');
  
  if (lines.length === 0) {
    process.stdout.write('[]');
    return;
  }

  const headers = lines[0].split(',').map(h => h.trim());
  const result: OutputPerson[] = [];

  for (let i = 1; i < lines.length; i++) {
    const row = lines[i].split(',');
    
    // Map header indices
    const idx = (field: string) => headers.indexOf(field);
    
    const name = row[idx('Name')];
    const birthdayStr = row[idx('Birthday')];
    const father = row[idx('Father')];
    const mother = row[idx('Mother')];
    const brother = row[idx('Brother')];
    const sister = row[idx('Sister')];

    const personName = parseName(name);
    if (!personName) continue;

    const birthday = parseDate(birthdayStr);
    if (!birthday) continue;

    // Format Birthday as YYYY-MM-DD
    const year = birthday.getFullYear();
    const month = String(birthday.getMonth() + 1).padStart(2, '0');
    const day = String(birthday.getDate()).padStart(2, '0');
    const formattedBirthday = `${year}-${month}-${day}`;

    const age = calculateAge(birthday, REF_DATE);

    const relatives: Relative[] = [];

    const addRelative = (nameStr: string, relationship: string) => {
      const parsed = parseName(nameStr);
      if (parsed) {
        relatives.push({
          FirstName: parsed.first,
          LastName: parsed.last,
          Relationship: relationship
        });
      }
    };

    addRelative(father, 'Father');
    addRelative(mother, 'Mother');
    addRelative(brother, 'Brother');
    addRelative(sister, 'Sister');

    result.push({
      FirstName: personName.first,
      LastName: personName.last,
      Birthday: formattedBirthday,
      Age: age,
      Relatives: relatives
    });
  }

  process.stdout.write(JSON.stringify(result, null, 2));
}

main();

File: codegen/results/v002/qwen3.5_27b_nothink/20260411_193102/go/007_beatles_interview/solution.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

type Relative struct {
	FirstName    string `json:"FirstName"`
	LastName     string `json:"LastName"`
	Relationship string `json:"Relationship"`
}

type Person struct {
	FirstName string     `json:"FirstName"`
	LastName  string     `json:"LastName"`
	Birthday  string     `json:"Birthday"`
	Age       int        `json:"Age"`
	Relatives []Relative `json:"Relatives"`
}

func parseName(fullName string) (string, string) {
	parts := strings.Fields(fullName)
	if len(parts) == 0 {
		return "", ""
	}
	firstName := parts[0]
	lastName := ""
	if len(parts) > 1 {
		lastName = parts[len(parts)-1]
	}
	return firstName, lastName
}

func parseRelative(fullName, relationship string) *Relative {
	if fullName == "" || fullName == "null" || fullName == "NULL" {
		return nil
	}
	firstName, lastName := parseName(fullName)
	return &Relative{
		FirstName:    firstName,
		LastName:     lastName,
		Relationship: relationship,
	}
}

func calculateAge(birthdayStr string) int {
	layout := "1/2/2006"
	birthday, err := time.Parse(layout, birthdayStr)
	if err != nil {
		return 0
	}

	refDate := time.Date(2025, time.July, 1, 0, 0, 0, 0, time.UTC)
	age := refDate.Year() - birthday.Year()

	// Adjust if birthday hasn't occurred yet in the reference year
	if refDate.YearDay() < birthday.YearDay() {
		age--
	}
	
	// Handle edge case where Month/Day logic needs more precision than YearDay
	// YearDay is 0-365, but let's double check month/day specifically
	if age < 0 {
		age = 0
	}
	
	// More precise age calculation
	anniversary := time.Date(refDate.Year(), birthday.Month(), birthday.Day(), 0, 0, 0, 0, time.UTC)
	if refDate.Before(anniversary) {
		age--
	}

	return age
}

func formatBirthdate(birthdayStr string) string {
	// Input format: M/D/YYYY or MM/D/YYYY or M/DD/YYYY
	// Target format: YYYY-MM-DD
	
	// We need to parse the specific input format
	// The input format doesn't have leading zeros consistently
	layout := "1/2/2006"
	t, err := time.Parse(layout, birthdayStr)
	if err != nil {
		return birthdayStr
	}
	return t.Format("2006-01-02")
}

func main() {
	file, err := os.Open("input/input.csv")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error opening file: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading CSV: %v\n", err)
		os.Exit(1)
	}

	if len(records) < 2 {
		fmt.Println("[]")
		return
	}

	// Skip header
	dataRows := records[1:]
	var people []Person

	for _, row := range dataRows {
		if len(row) < 7 {
			continue
		}

		name := row[0]
		birthdayRaw := row[1]
		father := row[3]
		mother := row[4]
		brother := row[5]
		sister := row[6]

		firstName, lastName := parseName(name)
		birthdayFormatted := formatBirthdate(birthdayRaw)
		age := calculateAge(birthdayRaw)

		var relatives []Relative
		if rel := parseRelative(father, "Father"); rel != nil {
			relatives = append(relatives, *rel)
		}
		if rel := parseRelative(mother, "Mother"); rel != nil {
			relatives = append(relatives, *rel)
		}
		if rel := parseRelative(brother, "Brother"); rel != nil {
			relatives = append(relatives, *rel)
		}
		if rel := parseRelative(sister, "Sister"); rel != nil {
			relatives = append(relatives, *rel)
		}

		person := Person{
			FirstName: firstName,
			LastName:  lastName,
			Birthday:  birthdayFormatted,
			Age:       age,
			Relatives: relatives,
		}
		people = append(people, person)
	}

	encoder := json.NewEncoder(os.Stdout)
	encoder.SetIndent("", "  ")
	if err := encoder.Encode(people); err != nil {
		fmt.Fprintf(os.Stderr, "Error encoding JSON: %v\n", err)
		os.Exit(1)
	}
}